### PR TITLE
LLVM CI: build LLVM from source

### PR DESCRIPTION
**Pull request by HertzDevil**
The official Apt repositories only serve the 3 most recent major versions, while the volunteer-built packages are incomplete and of inconsistent quality, so the best solution is to build LLVM ourselves.

This also reduces the total size for the 6 LLVM caches from 8 GB to around 2.4 GB. That's more than half the limit for a GitHub repository.

`lld` is needed for #13193; technically speaking, we could build it straight from the very repository we're cloning here, but we aren't really testing any version-specific `lld` behavior, so this patch grabs it from Apt.

### CODE REVIEW

**Comment by oprypin** [comment]
> This also reduces the total size for the 6 LLVM caches

Hmm actually it looks like there wasn't even any reason to have those caches before because it just moves the download+extract step from one place on GitHub to another.
Now of course they'll be necessary. 
Looks good

File: .github/workflows/llvm.yml [hunk 1]
```diff
@@ -38,19 +33,30 @@ jobs:
         uses: actions/cache@v4
         with:
           path: ./llvm
-          key: llvm-${{ matrix.llvm_version }}
+          key: llvm-libs-${{ matrix.llvm_version }}-${{ hashFiles('.github/workflows/llvm.yml') }}-linux
```
Comment (straight-shoota) [inline .github/workflows/llvm.yml:36]: **issue:** The hash on this workflow file triggers a rebuild every time we change anything here. This includes every time we bump the version for `install-crystal` action after each release. I think that's quite frequent in relation to the long build times for LLVM.
I'm wondering if we could narrow this down somehow, maybe call a separate workflow which handles only building LLVM?

**Comment by straight-shoota** [comment]
I was actually about to look into using `nix` more for our builds.

We already have an environment with configurable LLVM versions in `shell.nix`. The dependencies are defined and cached in nixpkgs. And `nixpkgs_unstable` currently has package definitions for LLVM 8 through LLVM 18. So we don't need to build anything ourselves

**Comment by ysbaddaden** [comment]
> The official Apt repositories only serve the 3 most recent major versions

You mean the official Debian or Ubuntu APT repositories?

Because the official [LLVM APT repositories](https://apt.llvm.org/) have all the releases we want. For example LLVM 13 to 18 for Ubuntu 22.04 + nightlies of LLVM 19 (see https://apt.llvm.org/jammy/pool/main/l/), and it will continue to receive the new releases for years to come. For Ubuntu 20.04 they have all releases since LLVM 9 (see https://apt.llvm.org/focal/pool/main/l/).

**Comment by HertzDevil** [comment]
Noble, the next Ubuntu version that is going to become LTS, [serves only 17 and beyond](https://apt.llvm.org/noble/pool/main/l/), so we risk being unable to update to a more recent Ubuntu runner image. We cannot really drop CI for LLVM 13 yet, and technically all the way down to LLVM 11, the most recent version released when Crystal 1.0 was out.

**Pull request closed without merging**